Remove classesDirs dependency on DefaultSourceSetOutput (#22570)
gradle/gradle: subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc

For some examples of possible error or warning outputs which may arise due to this change, see <<remove_test_add_opens>>.

==== SourceSet classesDirs no longer depends upon the entire SourceSet as a task dependency

> Prior to Gradle 8.0, the task dependencies for the `classesDirs` of a given `SourceSetOutput`

Reviewer: Can this link to the method in the DSL ref or Javadoc?

> Prior to Gradle 8.0, the task dependencies for the `classesDirs` of a given `SourceSetOutput`
> included the `SourceSetOutput` itself as a dependency. This meant that directly
> computing `classesDirs` would also execute `classes`, `processResources`, and any other
> task dependency for the `SourceSetOutput`.
> Since 8.0, this implicit dependency is removed. Now, computing `classesDirs` only executes the
> tasks which directly contribute to it and none of the tasks which contribute to the parent `SourceSetOutput`.

Reviewer: ```suggestion
Prior to Gradle 8.0, the task dependencies for the `SourceSetOutput.classesDirs` 
included tasks that did not produce class files. This meant that a task that
depends on `classesDirs` would also depend on `classes`, `processResources`, and any other
task dependency added to `SourceSetOutput`. This behavior was potentially an error because the `classesDirs` property did not contain the output for `processResources`. 
Since 8.0, this implicit dependency is removed. Now, depending on `classesDirs` only executes the
tasks which directly produce files in the classes directories.
```

Consider the following buildscript:
```groovy
plugins {
    id 'java-library'
}

> def fooTask = tasks.register("foo") {
>     outputs.file(project.layout.buildDirectory.dir("fooOut"))
> }
> tasks.register("bar") {
>     inputs.files(java.sourceSets.main.output.classesDirs)
> }
> java {
>     sourceSets {
>         main {
>             resources.srcDir(fooTask.map { it.outputs.files.singleFile })
>         }
>     }
> }

Reviewer: I think I would focus this on just the new behavior of classesDirs and with a concrete kind of example.  It's a little disconnected that the fooTask isn't executed. It's not executed because processResources isn't executed.

```suggestion
// Task lists all files in the given classFiles FileCollection
tasks.register("listClassFiles", ListClassFiles) {
    classFiles.from(java.sourceSets.main.output.classesDirs)
}
```

```

> Previously, the `bar` task included `compileJava`, `foo`, `processResources`, and `classes` as task
> dependencies. Now, only `compileJava` is a task dependency of `bar`.

Reviewer: ```suggestion
Previously, the `listClassFiles` task depended on `compileJava`, `processResources`, and `classes`. Now, only `compileJava` is a task dependency of `listClassFiles`.
```

> To mitigate this change, any task which previously relied on computing all source set outputs
> should instead depend on the parent `SourceSetOutput`, which is itself a `FileCollection`
> containing all classes and resources.
> 
> If depending on the `SourceSetOutput` directly is not feasible, the following snippet will
> restore the previous behavior:
> ```groovy
> java {
>     sourceSets {
>         main {
>             output.classesDirs.builtBy(output)
>         }
>     }
> }
> ```

Reviewer: ```suggestion
If a task in your build relied on the previous behavior, you can instead use the entire `SourceSetOutput` as an input, which contains all classes and resources.

If that is not feasible, you can restore the previous behavior by adding more task dependencies to `classesDirs`:
```groovy
java {
    sourceSets {
        main {
            output.classesDirs.builtBy(output)
        }
    }
}
```
```
